File: adventofcode/2021/3/python3/3.py
def commonty(l, most_common=True, untie='0'):
    c0 = l.count('0')
    c1 = l.count('1')
    if c0 > c1: return '0' if most_common else '1'
    if c0 < c1: return '1' if most_common else '0'
    return untie

def slv(binaries):
    a = "".join([commonty([binary[i] for binary in binaries]) for i in range(len(binaries[0]))])
    b = a.replace('0', '2').replace('1', '0').replace('2', '1')
    return int(a, 2) * int(b, 2)

def slv2(binaries, most_common, untie):
    prefix_mask = ''
    while len(binaries) > 1:
        prefix_mask = prefix_mask + commonty([binary[len(prefix_mask)] for binary in binaries], most_common, untie)
        binaries = [binary for binary in binaries if binary.startswith(prefix_mask)]
    return int(binaries[0], 2)

binaries = []
while True:
    try:
        binaries.append(input().rstrip())
    except EOFError: break
print("Problem1: {}".format(slv(binaries)))
print("Problem2: {}".format(slv2(binaries.copy(), most_common=True, untie='1') * slv2(binaries.copy(), most_common=False, untie='0')))

File: adventofcode/2021/3/kotlin/3.kt
fun commonty(bits: List<Char>, mostCommon : Boolean = true, untie: Char = '0') =
    bits.partition {
        bit -> bit == '0'
    }.let{
        (zeros, ones) -> when(zeros.size.compareTo(ones.size)) {
            1 -> if (mostCommon) '0' else '1'
            -1 -> if (mostCommon) '1' else '0'
            else -> untie
        }
    }

fun slv(binaries: List<String>, m: Int): Int {
    val mostCommon = (0 until m).map { index -> binaries.map { it[index] }.let { commonty(it) } }.joinToString("")
    val leastCommon = mostCommon.map {
        bit -> when(bit) {
            '0' -> '1'
            else -> '0'
        }
    }.joinToString("")
    return Integer.parseInt(mostCommon, 2) * Integer.parseInt(leastCommon, 2)
}

fun slv2(binaries: List<String>, mostCommon: Boolean, untie: Char): Int {
    var binariesAux: List<String> = binaries
    var prefix = ""
    while (binariesAux.size > 1) {
        prefix += commonty(binariesAux.map { it[prefix.length] }, mostCommon, untie)
        binariesAux = binariesAux.filter { binary -> binary.startsWith(prefix) }
    }
    return Integer.parseInt(binariesAux[0], 2)
}

fun main() {
    val (n, m) = readLine()!!.split(" ").map(String::toInt)
    val binaries = (0 until n).map { readLine()!! }
    println("Problem 1: ${slv(binaries, m)}")
    println("Problem 2: ${slv2(binaries, true, '1') * slv2(binaries,false, '0')}")
}